sw/airborne/modules/ctrl/optical_flow_landing.c

```diff
@@ -41,127 +41,151 @@
 #endif
-#ifndef OPTICAL_FLOW_LANDING_VISION_METHOD
-#define OPTICAL_FLOW_LANDING_VISION_METHOD 1
+
+#ifndef OFL_VISION_METHOD
+#define OFL_VISION_METHOD 1
 #endif
-#ifndef OPTICAL_FLOW_LANDING_CONTROL_METHOD
-#define OPTICAL_FLOW_LANDING_CONTROL_METHOD 0
+
+#ifndef OFL_CONTROL_METHOD
+#define OFL_CONTROL_METHOD 0
 #endif
-#ifndef OPTICAL_FLOW_LANDING_COV_METHOD
-#define OPTICAL_FLOW_LANDING_COV_METHOD 0
+
+#ifndef OFL_COV_METHOD
+#define OFL_COV_METHOD 0
 #endif
+
 // number of time steps used for calculating the covariance (oscillations)
-#ifndef OPTICAL_FLOW_LANDING_COV_WINDOW_SIZE
-#define OPTICAL_FLOW_LANDING_COV_WINDOW_SIZE 30
+#ifndef OFL_COV_WINDOW_SIZE
+#define OFL_COV_WINDOW_SIZE 30
+#endif
+
+#ifndef OFL_COV_LANDING_LIMIT
+#define OFL_COV_LANDING_LIMIT 2.2
+#endif
+
+#ifndef OFL_COV_SETPOINT
+#define OFL_COV_SETPOINT -0.0075
+#endif
+
+#ifndef OFL_LP_CONST
+#define OFL_LP_CONST 0.02
 #endif
 
+#ifndef OFL_P_LAND_THRESHOLD
+#define OFL_P_LAND_THRESHOLD 0.15
+#endif
+
+#ifndef OFL_ELC_OSCILLATE
+#define OFL_ELC_OSCILLATE true
+#endif
+
+// Constants
+// minimum value of the P-gain for divergence control
+// adaptive control / exponential gain control will not be able to go lower
+#define MINIMUM_GAIN 0.1
+
+// for exponential gain landing, gain increase per second during the first (hover) phase:
+#define INCREASE_GAIN_PER_SECOND 0.02
+
 // variables retained between module calls
-float thrust_history[MAX_COV_WINDOW_SIZE];
-float divergence_history[MAX_COV_WINDOW_SIZE];
-float past_divergence_history[MAX_COV_WINDOW_SIZE];
-float dt_history[MAX_COV_WINDOW_SIZE];
-unsigned long ind_hist;
-float divergence;
 float divergence_vision;
 float divergence_vision_dt;
 float normalized_thrust;
 float cov_div;
-float pstate;
-float pused;
+float pstate, pused;
 float istate;
 float dstate;
-float dt;
-int vision_message_nr;
-int previous_message_nr;
-int landing;
-float previous_err;
+float vision_time,  prev_vision_time;
+bool landing;
 float previous_cov_err;
+int32_t thrust_set;
+float divergence_setpoint;
+
 // for the exponentially decreasing gain strategy:
-int elc_phase;
+int32_t elc_phase;
 uint32_t elc_time_start;
-float elc_p_gain_start;
-float elc_i_gain_start;
-float elc_d_gain_start;
-long count_covdiv;
+float elc_p_gain_start, elc_i_gain_start,  elc_d_gain_start;
+int32_t count_covdiv;
 float lp_cov_div;
 
-uint32_t previous_time;
-uint32_t module_enter_time;
+static abi_event agl_ev; ///< The altitude ABI event
+static abi_event optical_flow_ev;
+
+// struct containing most relevant parameters
+struct OpticalFlowLanding of_landing_ctrl;
 
 // sending the divergence message to the ground station:
 static void send_divergence(struct transport_tx *trans, struct link_device *dev)
 {
   pprz_msg_send_DIVERGENCE(trans, dev, AC_ID,
-                           &divergence, &divergence_vision_dt, &normalized_thrust,
+                           &(of_landing_ctrl.divergence), &divergence_vision_dt, &of_landing_ctrl.nominal_thrust,
```

Comment:
Nominal thrust was supposed to be a constant, while normalized thrust would be between 0 (minimum thrust available to the controller) and 1 (maximum thrust allowed) - which is hence varying during control. I would prefer normalized thrust here. 

divergence is low-passed. Is of_landing_ctrl.divergence as well? We don't want copies of the divergence from the vision, we want the divergence that is used by control.